### PR TITLE
Fix for backend service update

### PR DESCRIPTION
**Pull request by kl52752**
We should update backend service with ig llinks
only when ig links changed.

### CODE REVIEW

**Review by cezarygerard — APPROVED** [review]
lgtm

File: pkg/backends/regional_ig_linker.go [hunk 1]
```diff
@@ -46,30 +45,26 @@ func (linker *RegionalInstanceGroupLinker) Link(sp utils.ServicePort, projectID
...
 	}
-
-	addIGs := sets.String{}
-	for _, igLink := range igLinks {
-		path, err := utils.RelativeResourceName(igLink)
-		if err != nil {
-			return fmt.Errorf("failed to parse instance group %s: %w", igLink, err)
-		}
-		addIGs.Insert(path)
+	be, err := linker.backendPool.Get(sp.BackendName(), meta.VersionGA, meta.Regional)
```
Comment (panslava) [inline pkg/backends/regional_ig_linker.go:48]: for me it is not very clear, what stands behind `be`, is it `BackEndservice` ?

Comment (panslava) [reply pkg/backends/regional_ig_linker.go:48]: (oh, it is probably coming from ig_linker.go)

**Review by panslava — APPROVED** [review]

**Comment by cezarygerard** [comment]
@bowei ping  we need this PR for RBS launch

**Comment by bowei** [comment]
Let's fix the commit message so that the change is obvious when looking at the git history:

```
Only update BackendService IG links when they change

<more details...>
```

File: pkg/backends/regional_ig_linker.go [hunk 2]
```diff
@@ -46,33 +45,29 @@ func (linker *RegionalInstanceGroupLinker) Link(sp utils.ServicePort, projectID
...
+		return err
 	}
 	if len(addIGs) == 0 {
 		return nil
 	}
-	// TODO(kl52752) Check for existing links and add only new one
+
 	var newBackends []*composite.Backend
-	for _, igLink := range addIGs.List() {
+	for _, igLink := range addIGs {
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:61]: Is it an issue for unit tests that this order may not be stable due to the use of a map in the sets.String? Should the tests account for that?

File: pkg/backends/regional_ig_linker.go [hunk 3]
```diff
@@ -46,33 +45,29 @@ func (linker *RegionalInstanceGroupLinker) Link(sp utils.ServicePort, projectID
...
+	bs, err := linker.backendPool.Get(sp.BackendName(), meta.VersionGA, meta.Regional)
+	if err != nil {
+		return err
+	}
+	addIGs, err := getInstanceGroupsToAdd(bs, igLinks)
+	if err != nil {
+		return err
 	}
 	if len(addIGs) == 0 {
 		return nil
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:57]: Add a log here that there was no change to the Backends.

File: pkg/backends/regional_ig_linker.go [hunk 4]
```diff
@@ -46,33 +45,29 @@ func (linker *RegionalInstanceGroupLinker) Link(sp utils.ServicePort, projectID
 		igSelfLink := cloudprovider.SelfLink(meta.VersionGA, projectID, "instanceGroups", key)
 		igLinks = append(igLinks, igSelfLink)
 	}
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:47]: Let's log at the beginning of this method (Link()) 

log.Infof("Link(%v, %q, %v)", sp, projectID, zones)

File: pkg/backends/regional_ig_linker.go [hunk 5]
```diff
@@ -46,33 +45,29 @@ func (linker *RegionalInstanceGroupLinker) Link(sp utils.ServicePort, projectID
...
-	be, err := linker.backendPool.Get(sp.BackendName(), meta.VersionGA, meta.Regional)
-	if err != nil {
-		return err
-	}
-	be.Backends = newBackends
 
-	if err := linker.backendPool.Update(be); err != nil {
+	bs.Backends = newBackends
+
+	if err := linker.backendPool.Update(bs); err != nil {
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:70]: let's log Updating backendPool

**Review by bowei — COMMENTED** [review]
Mostly LGTM. Let's add logging so if there is an issue with this code path, we have enough information to debug.

File: pkg/backends/regional_ig_linker.go (hunk 3 above)
Comment (kl52752) [reply pkg/backends/regional_ig_linker.go:57]: done

File: pkg/backends/regional_ig_linker.go (hunk 5 above)
Comment (kl52752) [reply pkg/backends/regional_ig_linker.go:70]: done

File: pkg/backends/regional_ig_linker.go (hunk 4 above)
Comment (kl52752) [reply pkg/backends/regional_ig_linker.go:47]: done

File: pkg/backends/regional_ig_linker.go (hunk 2 above)
Comment (kl52752) [reply pkg/backends/regional_ig_linker.go:64]: Added UT

File: pkg/backends/regional_ig_linker.go [hunk 6]
```diff
@@ -39,40 +39,38 @@ func NewRegionalInstanceGroupLinker(instancePool instances.NodePool, backendPool
...
+	bs, err := linker.backendPool.Get(sp.BackendName(), meta.VersionGA, meta.Regional)
+	if err != nil {
+		return err
+	}
+	addIGs, err := getInstanceGroupsToAdd(bs, igLinks)
+	if err != nil {
+		return err
 	}
 	if len(addIGs) == 0 {
+		klog.V(3).Infof("No backends to add for %s, skipping update.", sp.BackendName())
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:59]: Can we have this at V(2) so it appears in the normal log? Is it too noisy?

Comment (bowei) [reply pkg/backends/regional_ig_linker.go:59]: We can do this as a follow up

File: pkg/backends/regional_ig_linker.go [hunk 7]
```diff
@@ -39,40 +39,38 @@ func NewRegionalInstanceGroupLinker(instancePool instances.NodePool, backendPool
...
 	}
-	be, err := linker.backendPool.Get(sp.BackendName(), meta.VersionGA, meta.Regional)
-	if err != nil {
-		return err
-	}
-	be.Backends = newBackends
 
-	if err := linker.backendPool.Update(be); err != nil {
+	bs.Backends = newBackends
+	klog.V(3).Infof("Update Backend %s, with %d backends.", sp.BackendName(), len(addIGs))
```
Comment (bowei) [inline pkg/backends/regional_ig_linker.go:72]: Same comment here V(2)

**Pull request merged**